execution-engine/src/hcall/buffer.rs

```diff
@@ -0,0 +1,138 @@
+//! The Veracruz host-call interface, and its implementation.
+//!
+//! ## Authors
+//!
+//! The Veracruz Development Team.
+//!
+//! ## Copyright
+//!
+//! See the file `LICENSE.markdown` in the Veracruz root directory for licensing
+//! and copyright information.
+//!
+///! GLUE CODE FOR NEW API.
+///! IT SHOULD BE REPLACED BY VFS IMPLEMENTATION OF WASI API.
+
+use err_derive::Error;
+use std::{collections::HashMap, result::Result, vec::Vec, string::{ToString, String}};
+use veracruz_utils::{VeracruzCapabilityIndex, VeracruzCapability, VeracruzCapabilityTable};
+use serde::{Deserialize, Serialize};
+
+/// Error type for mexico-city buffer.
+#[derive(Clone, Debug, Error, Serialize, Deserialize)]
+pub enum VFSError {
+    //TODO: potential remove this 
+    #[error(
+        display = "VFSError: File {} cannot be found.", _0
+    )]
+    FileNotFound(String),
+    #[error(
+        display = "VFSError: Principal or program {:?} cannot be found.",_0
+    )]
+    IndexNotFound(VeracruzCapabilityIndex),
+    #[error(
+        display = "VFSError: Client {:?} is disallowed to {:?}.",client_id,operation
+    )]
+    CapabilityDenial {
+        client_id: VeracruzCapabilityIndex,
+        operation : VeracruzCapability,
+    },
+    #[error(
+        display = "VFSError: File {:?} digest mismatches, model: {:?}, received: {:?}.",file_name, model, received
+    )]
+    FileDigestMismatch{
+        file_name : String,
+        model : Vec<u8>,
+        received : Vec<u8>,
+    },
+}
+
+/// Buffer for storing program, (static) data, and stream data.
+#[derive(Clone, Debug)]
+pub struct VFS {
+    fs: HashMap<String, Vec<u8>>,
+    capabilities: VeracruzCapabilityTable,
+    digests: HashMap<String, Vec<u8>>,
+}
+
+impl VFS {
+    /// Computes a SHA-256 digest of the bytes passed to it in `buffer`.
+    fn sha_256_digest(buffer: &[u8]) -> Vec<u8> {
+        ring::digest::digest(&ring::digest::SHA256, buffer)
+            .as_ref()
+            .to_vec()
+    }
+
+    /// Initialize a new empty buffer.
+    pub fn new(capabilities: &VeracruzCapabilityTable, digests: &HashMap<String, Vec<u8>>) -> Self {
+        VFS {
+            fs: HashMap::new(),
+            capabilities: capabilities.clone(),
+            digests: digests.clone(),
+        }
+    }
+    
+    pub fn write(&mut self, file_name : &str, new_data : &[u8]) -> Result<(),VFSError> {
```

Comment:
Missing Rustdoc comments on these functions (many more occurrences below).

Reply:
This file is glue code, and will be removed in wasi change. I have documented it in the beginning in case I forgot.